src/overlay/Peer.cpp

```diff
@@ -630,38 +710,256 @@ Peer::recvMessage(StellarMessage const& stellarMsg)
+            auto minSlotToRemember =
+                mApp.getHerder().getMinLedgerSeqToRemember();
+            if (slotIndex >= minSlotToRemember &&
+                mApp.getHerder().isLatestSCPMessage(msg.mMessage))
+            {
+                return false;
+            }
+            // Trim an SCP message if it's outside of range or we have a newer
+            // statement
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    };
+
+    int dropped = 0;
+    auto it = mFlowControlledQueue.cbegin();
+    while (it != mFlowControlledQueue.cend())
+    {
+        if (shouldTrim(*it))
+        {
+            getOutboundFloodTimer(mApp, it->mMessage)
+                .Update(now - it->mTimeEmplaced);
+            it = mFlowControlledQueue.erase(it);
+            dropped++;
+        }
+        else
+        {
+            ++it;
+        }
+    }
+    if (dropped)
+    {
+        CLOG_TRACE(Overlay, "Dropped {} messages to peer {}", dropped,
+                   mApp.getConfig().toShortString(getPeerID()));
+    }
+}
+
+void
+Peer::maybeShedLoadAndSendNextBatch()
+{
+    releaseAssert(mRemoteOverlayVersion >=
+                  Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL);
+    releaseAssert(mRemotePeerWantsFlowControl && *mRemotePeerWantsFlowControl);
+
+    // First see if we can drop some messages
+    maybeTrimQueue();
+
+    auto oldOutboundCapacity = mOutboundCapacity;
+    auto it = mFlowControlledQueue.begin();
+    while (it != mFlowControlledQueue.end() && mOutboundCapacity > 0)
+    {
+        sendAuthenticatedMessage(it->mMessage);
+        getOutboundFloodTimer(mApp, it->mMessage)
+            .Update(mApp.getClock().now() - it->mTimeEmplaced);
+
+        // Only count flood messages towards capacity
+        if (mApp.getOverlayManager().isFloodMessage(it->mMessage))
+        {
+            mOutboundCapacity--;
+        }
+        it = mFlowControlledQueue.erase(it);
+    }
+
+    CLOG_TRACE(Overlay, "Peer {}: send next flood batch of {}",
+               mApp.getConfig().toShortString(getPeerID()),
+               (oldOutboundCapacity - mOutboundCapacity));
+}
+
+void
+Peer::messageProcessed(StellarMessage const& msg)
+{
+    bool flowControl =
+        mRemoteOverlayVersion >= Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL &&
+        mApp.getConfig().ENABLE_OVERLAY_FLOW_CONTROL;
+    if (shouldAbort() || !flowControl)
+    {
+        return;
+    }
+
+    mCapacity.mTotalCapacity++;
+    if (mApp.getOverlayManager().isFloodMessage(msg))
+    {
+        mCapacity.mFloodCapacity++;
+        CLOG_TRACE(Overlay, "{}: got capacity {} for peer {}",
+                   mApp.getConfig().toShortString(
+                       mApp.getConfig().NODE_SEED.getPublicKey()),
+                   mCapacity.mFloodCapacity,
+                   mApp.getConfig().toShortString(getPeerID()));
+        mFloodMsgsProcessed++;
+
+        auto batchFraction = std::max<uint32_t>(
```

Comment:
this should be called `batchSize`... actually this is the only place where `FLOW_CONTROL_SEND_MORE_PIPELINE_RATIO` is called, so maybe the setting should just be `FLOW_CONTROL_SEND_MORE_BATCH_SIZE` (as I am not even sure it's that meaningful anyways as a fraction)